datastore/longhorn.go

```diff
@@ -233,6 +232,59 @@ func (s *DataStore) ListSettings() (map[types.SettingName]*longhorn.Setting, err
 	return itemMap, nil
 }
 
+// AnnotateAWSIAMRoleArn ensures that the running pods of the manager as well as the replica instance managers.
+// have the correct AWS IAM role arn assigned to them based on the passed `awsIAMRoleArn`
+func (s *DataStore) AnnotateAWSIAMRoleArn(controllerID, awsIAMRoleArn string) (bool, error) {
+	update := false
+
+	selector, err := s.getManagerSelector()
+	if err != nil {
+		return update, err
+	}
+	managerPods, err := s.pLister.Pods(s.namespace).List(selector)
+	if err != nil {
+		return update, err
+	}
+
+	selector, err = metav1.LabelSelectorAsSelector(&metav1.LabelSelector{
+		MatchLabels: types.GetInstanceManagerLabels("", "", types.InstanceManagerTypeReplica),
+	})
+	if err != nil {
+		return update, err
+	}
+	imPods, err := s.pLister.Pods(s.namespace).List(selector)
+	if err != nil {
+		return update, err
+	}
+
+	pods := append(managerPods, imPods...)
+	for _, pod := range pods {
+		if pod.Spec.NodeName != controllerID {
+			continue
+		}
+
+		val, exist := pod.Annotations[types.AWSIAMRroleAnnotation]
+		updateAnnotation := awsIAMRoleArn != "" && awsIAMRoleArn != val
+		deleteAnnotation := awsIAMRoleArn == "" && exist
+		if updateAnnotation {
+			if pod.Annotations == nil {
+				pod.Annotations = make(map[string]string)
+			}
+			pod.Annotations[types.AWSIAMRroleAnnotation] = awsIAMRoleArn
+		} else if deleteAnnotation {
+			delete(pod.Annotations, types.AWSIAMRroleAnnotation)
+		} else {
+			continue
+		}
+
+		update = update || true
```

Comment:
This will make the update successful if at least one pod updated successfully. Is this expected?

I think we need to return the error to let event requeued to try update again.

Reply:
It's a `update` variable to determine to print log `Annotates the AWS IAM role succeeded` or not.
If one of the Pod updates failed, the error returned immediately and the event would be re-queue.

```go
        update, err := ks.ds.AnnotateAWSIAMRoleArn(ks.controllerID, awsIAMRoleArn)
	if err != nil {
		return err
	}

	if update {
		ks.logger.Infof("Annotates the AWS IAM role succeeded")
	}
```

Reply:
Yes, but it's always `true` because of `update = update || true`, so I doubt the `if update` any meaning in this pod update loop?

Reply:
After checking again, there is a case for non-update, so it does have some meaning :)